server/sonar-process-monitor/src/main/java/org/sonar/application/process/ProcessLauncherImpl.java

```diff
@@ -0,0 +1,256 @@
+
+import static java.lang.String.format;
+import static org.sonar.process.ProcessEntryPoint.PROPERTY_PROCESS_INDEX;
+import static org.sonar.process.ProcessEntryPoint.PROPERTY_PROCESS_KEY;
+import static org.sonar.process.ProcessEntryPoint.PROPERTY_SHARED_PATH;
+import static org.sonar.process.ProcessEntryPoint.PROPERTY_TERMINATION_TIMEOUT;
+
+public class ProcessLauncherImpl implements ProcessLauncher {
+  private static final Logger LOG = LoggerFactory.getLogger(ProcessLauncherImpl.class);
+
+  private final File tempDir;
+  private final AllProcessesCommands allProcessesCommands;
+  private final Supplier<ProcessBuilder> processBuilderSupplier;
+
+  public ProcessLauncherImpl(File tempDir) {
+    this(tempDir, new AllProcessesCommands(tempDir), JavaLangProcessBuilder::new);
+  }
+
+  ProcessLauncherImpl(File tempDir, AllProcessesCommands allProcessesCommands, Supplier<ProcessBuilder> processBuilderSupplier) {
+    this.tempDir = tempDir;
+    this.allProcessesCommands = allProcessesCommands;
+    this.processBuilderSupplier = processBuilderSupplier;
+  }
+
+  @Override
+  public void close() {
+    allProcessesCommands.close();
+  }
+
+  @Override
+  public ProcessMonitor launch(EsCommand esCommand) {
+    Process process = null;
+    try {
+      writeConfFiles(esCommand);
+      ProcessBuilder processBuilder = create(esCommand);
+      LOG.info("Launch process[{}]: {}", esCommand.getProcessId().getKey(), String.join(" ", processBuilder.command()));
+
+      process = processBuilder.start();
+
+      return new EsProcessMonitor(process, esCommand);
+    } catch (Exception e) {
+      // just in case
+      if (process != null) {
+        process.destroyForcibly();
+      }
+      throw new IllegalStateException(format("Fail to launch process [%s]", esCommand.getProcessId().getKey()), e);
+    }
+  }
+
+  private void writeConfFiles(EsCommand esCommand) {
+    File confDir = esCommand.getConfDir();
+    if (!confDir.exists() && !confDir.mkdirs()) {
+      String error = format("Failed to create temporary configuration directory [%s]", confDir.getAbsolutePath());
+      LOG.error(error);
+      throw new IllegalStateException(error);
+    }
+
+    try {
+      IOUtils.copy(getClass().getResourceAsStream("elasticsearch.yml"), new FileOutputStream(new File(confDir, "elasticsearch.yml")));
+      IOUtils.copy(getClass().getResourceAsStream("jvm.options"), new FileOutputStream(new File(confDir, "jvm.options")));
+      esCommand.getLog4j2Properties().store(new FileOutputStream(new File(confDir, "log4j2.properties")), "log42 properties file for ES bundled in SonarQube");
+    } catch (IOException e) {
+      e.printStackTrace();
+      throw new IllegalStateException("Failed to write ES configuration files", e);
+    }
+  }
+
+  @Override
+  public ProcessMonitor launch(JavaCommand javaCommand) {
+    Process process = null;
+    try {
+      ProcessCommands commands = allProcessesCommands.createAfterClean(javaCommand.getProcessId().getIpcIndex());
```

Comment:
![BLOCKER](https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER') Use try-with-resources or close this "ProcessCommands" in a "finally" clause. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2095)